operators.js

```diff
@@ -330,56 +330,86 @@ function asOffsets() {
 //#endregion
 //#region "Consumer operators": they execute the query tree
 
-async function executeDeferredOps(ops, meta) {
+function executeDeferredOps(ops, meta) {
   // Collect all deferred tasks and their object-traversal paths
   const allDeferred = []
   traverse.forEach(ops, function (val) {
     if (!val) return
     // this.block() means don't traverse inside these, they won't have DEFERRED
     if (this.key === 'meta' && val.jitdb) return this.block()
-    if (val.type === 'DEFERRED' && val.task) allDeferred.push([this.path, val])
+    if (val.type === 'DEFERRED' && val.task) {
+      allDeferred.push([this.path, val.task])
+    }
     if (!(Array.isArray(val) || val.type === 'AND' || val.type === 'OR')) {
       this.block()
     }
   })
-  if (allDeferred.length === 0) return ops
-
-  // Execute all deferred tasks and collect the results (and the paths)
-  const allResults = await Promise.all(
-    allDeferred.map(([path, obj]) =>
-      promisify(obj.task)(meta).then((result) => {
-        if (!result) return [path, {}]
-        else if (typeof result === 'function') return [path, result()]
-        else return [path, result]
-      })
-    )
-  )
+  if (allDeferred.length === 0) return pull.values([ops])
+
+  // State needed throughout the execution of the `readable`
+  const done = multicb({ pluck: 1 })
+  let completed = false
+  const abortListeners = []
+  function addAbortListener(listener) {
+    abortListeners.push(listener)
+  }
 
-  // Replace all deferreds with their respective results
-  allResults.forEach(([path, result]) => {
-    result.meta = meta
-    if (path.length === 0) ops = result
-    else traverse.set(ops, path, result)
-  })
+  return function readable(end, cb) {
+    if (end) {
+      while (abortListeners.length) abortListeners.shift()()
```

Comment:
I remember I once ran into a performance problem with this pattern but I think it should be okay here because one has to explicitly add an abort handler to get this triggered.